### ISSUE TITLE
Failed regular expression searches can mark whole file as search result

### ISSUE THREAD

**Issue by Uhf7** [notepad-plus-plus/notepad-plus-plus#9565]
### Description of the Issue
If a regular expression search fails because of an internal stack overflow or a similar exception, the whole file is marked as search result.

### Steps to Reproduce the Issue
1. Open the text file contained in this zip: [BackslashK-Testfile.zip](https://github.com/notepad-plus-plus/notepad-plus-plus/files/6030665/BackslashK-Testfile.zip)
2. Search for the regular expression in line 2 of the test file: `abc(.|\R)*?\Kdef`, easiest with wrap around option on.
3. Repeat the search, easiest with F3.

### Expected Behavior
- All occurrences of `def`, which are preceded by an `abc`, are found step-by-step.
- _Edit 1:_ If the search fails, then it should not select the whole file.

### Actual Behavior
- _Edit 2:_ Not all occurrences of `def`, which are preceded by an `abc`, are found step-by-step.
- The `def` occurrences are found only as long as the distance between `abc` and `def` is shorter than 22000 to 64000 characters. An exact number cannot be given, because the internal error which occurs here is a stack overflow.
- If the distance between `abc` and `def` is too long, then the whole file is selected as search result. _Edit_ 3: This is dangerous, especially for replace operations.

### Debug Information
Notepad++ v7.9.3   (32-bit)
Build time : Feb 14 2021 - 03:12:31
Path : C:\Tools\Npp\Npp_793\notepad++.exe
Admin mode : OFF
Local Conf mode : ON
OS Name : Windows 10 Pro (64-bit) 
OS Version : 1607
OS Build : 14393.576
Current ANSI codepage : 1252
Plugins : mimeTools.dll NppConverter.dll NppExport.dll

**Comment by sasumner** [comment]
In the "Actual Behavior" there's a phrase which leads me to believe what is being described there is really "Expected Behavior":

* "... instead of treating the find as failed (without search result)"

But in the "Expected Behavior" it says:

* "All occurrences of ... are found step-by-step"

These seem like two contrasting desires for behavior.

I think the "all occurrences are found" is the desired outcome.
But...sometimes it is not possible due to technological limits.

If it is left at "not found" then what was searched for by the user could truly occur in the data, and we have misled him by saying the search was unsuccessful.

Wouldn't it be better to say something like "Your search may match your data, but due to an unexpected problem, we can neither confirm nor deny that"?  Way too wordy, of course. :-)

This has been a long-standing issue, and I consider #4761 to be sort of the "master" issue about it, but that's just my categorization of it.

While I was typing this I see a PR has been submitted which "treats the find as failed".
Again, not sure that is the best thing to do, although it is one way of handling it.

**Comment by Ekopalypse** [comment]
@Uhf7 - nice, but yes, I see the same problem - how is the user informed when such an error has occurred?
Saying something like "0 hits" is not right. And what happens if I do this with multiple files?
Does it break at the first error that occurred?

**Comment by Uhf7** [comment]
OK, I edited my description a little for clarification.

> not sure that is the best thing to do, although it is one way of handling it.

I think, almost everything is better than select the whole file (because of the replace-operation). So this would prevent this dangerous behavior immediately. Passing regex errors through Scintilla to Notepad++ is another desirable thing. I will have another look on it.

**Comment by sasumner** [comment]
>Passing regex errors through Scintilla to Notepad++ is another desirable thing. I will have another look on it.

This might be a reasonable place to start:  https://github.com/notepad-plus-plus/notepad-plus-plus/issues/4761#issuecomment-413838281

**Comment by sasumner** [comment]
@Uhf7 As you continue to work on this, please give this posted by @Ekopalypse some consideration:

>And what happens if I do this with multiple files? Does it break at the first error that occurred?

I suppose there can be situations where the engine problem can occur on a first file of a group of hundreds of files (in a _Find in Files_) and would not occur on any other file.  In this case, should the _Search results_ window reflect this type of failure somehow and produce usual output for the remaining files?

**Comment by Uhf7** [comment]
OK, that are too many tasks at once, I think.

One task is the transfer of an error message from `SciLexer.dll` to `Notepad++.exe`, another task is the reaction to this errors within `Notepad++.exe`.

What I would go first for is:
- Search result -1 for not found.
- Search result -2 for error while searching.
 Here, we should see a more detailed error message about what went wrong. In the moment, only one error message exists: "Find: Invalid regular expression". So there is the goal to transfer a real error message to Notepad++, although it may not be localizable.

Another thing (to be solved after that) is:
- The reaction to Search result -2 is not ideal in all places of the Notepad++ code.
 I didn't investigate it too deeply until now, but, for example, the Count function seems to treat -1 and -2 quite the same way...
 I don't know what happens in the moment, if a search result -2 occurs while processing a bunch of files.

**Comment by sasumner** [comment]
>In the moment, only one error message exists: "Find: Invalid regular expression"

See #6043

>the Count function seems to treat -1 and -2 quite the same way...

Shown well here, in that same issue:

https://github.com/notepad-plus-plus/notepad-plus-plus/issues/6043#issuecomment-520989262

**Comment by Uhf7** [comment]
@sasumner, Thank you for the links, so I know that the problem is lingering for some time now...

**Comment by sasumner** [comment]
The above commit will also close #4761
